catalog/app/containers/Admin/Buckets.tsx

```diff
@@ -1,39 +1,177 @@
+import * as IO from 'io-ts'
+import { formatValidationErrors } from 'io-ts-reporters'
+import * as Types from 'io-ts-types'
 import * as R from 'ramda'
 import * as React from 'react'
-import * as redux from 'react-redux'
-import { Redirect, useHistory } from 'react-router-dom'
-import * as RF from 'redux-form/es/immutable'
+import * as RF from 'react-final-form'
+import * as RRDom from 'react-router-dom'
+import * as urql from 'urql'
 import * as M from '@material-ui/core'
 
 import * as Pagination from 'components/Pagination'
-import * as AuthSelectors from 'containers/Auth/selectors'
 import * as Notifications from 'containers/Notifications'
+import * as Model from 'model'
 import * as APIConnector from 'utils/APIConnector'
-import * as BucketConfig from 'utils/BucketConfig'
-import * as Config from 'utils/Config'
 import Delay from 'utils/Delay'
 import * as Dialogs from 'utils/Dialogs'
+import { BaseError } from 'utils/error'
 import MetaTitle from 'utils/MetaTitle'
 import * as NamedRoutes from 'utils/NamedRoutes'
 import parseSearch from 'utils/parseSearch'
-import * as Cache from 'utils/ResourceCache'
 import { useTracker } from 'utils/tracking'
 import * as validators from 'utils/validators'
 
 import * as Form from './Form'
 import * as Table from './Table'
-import * as data from './data'
 
-// default icon as returned by the registry
-const DEFAULT_ICON = 'https://d1zvn9rasera71.cloudfront.net/q-128-square.png'
+// TODO: move reusable helpers out of here
+class ValidationError extends BaseError {
+  constructor(e: IO.Errors) {
+    const formatted = formatValidationErrors(e).join('\n')
+    super(`ValidationError\n${formatted}`)
+  }
+}
+
+function tryDecode<A extends any>(t: IO.Type<A, any, any>, i: unknown) {
+  return FP.function.pipe(
+    t.decode(i),
+    FP.either.fold(
+      (e) => {
+        throw new ValidationError(e)
+      },
+      (a) => a,
+    ),
+  )
+}
+
+function decode<T>(codec: IO.Type<T, any, any>) {
+  return (i: unknown) => tryDecode(codec, i)
+}
+
+const normalizeString = R.trim
+
+// TODO: toNullable combinator
+// function toNullable(codec, isNull) {
+//   // encode as null if isNull(value)
+// }
+
+const SNS_ARN_RE = /^arn:aws(-|\w)*:sns:(-|\w)*:\d*:\S+$/
 
 const DO_NOT_SUBSCRIBE_STR = 'DO_NOT_SUBSCRIBE'
 const DO_NOT_SUBSCRIBE_SYM = Symbol(DO_NOT_SUBSCRIBE_STR)
 
-const SNS_ARN_RE = /^arn:aws(-|\w)*:sns:(-|\w)*:\d*:\S+$/
+type SnsFormValue = string | typeof DO_NOT_SUBSCRIBE_SYM
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+const SnsFormValue = new IO.Type<SnsFormValue>(
+  'SnsFormValue',
+  (i): i is SnsFormValue => i === DO_NOT_SUBSCRIBE_SYM || IO.string.is(i),
+  (u, c) =>
+    u === DO_NOT_SUBSCRIBE_SYM || typeof u === 'string'
+      ? IO.success(u)
+      : IO.failure(u, c),
+  R.identity,
+)
+
+type FormSpec<Obj extends {}> = {
+  [K in keyof Obj]: (formValues: Record<string, unknown>) => Obj[K]
+}
+
+const editFormSpec: FormSpec<Model.BucketUpdateInput> = {
```

Comment:
Are you planning to move this kind of stuff (convert one data format to another) to Model?

Reply:
in this particular case -- no, because it's not reusable -- it converts a local data representation (form data) into more general / reusable / canonical format defined in the schema / model

Reply:
however, i do plan to find / write some harness for defining forms (something akin to https://github.com/ForbesLindesay/define-form)